### ISSUE TITLE
Arrow.write hangs on Tables.partitioner

### ISSUE THREAD

**Issue by kescobo** [JuliaData/Arrow.jl#108]
I've got a list of ~ 1200 CSV files that I'm trying to read into an arrow file. Reading them all into a DataFrame takes less than a second, and then writing that DataFrame to an arrow file also takes less than a second

```
julia> @time let df = DataFrame()
          for profile in profiles
              append!(df, CSV.File(profile, datarow=5, header=["clade", "NCBI_taxid", "abundance", "additional_species"]))
          end
          Arrow.write("test.arrow", df)
       end
  0.608143 seconds (670.79 k allocations: 172.351 MiB, 14.65% gc time)
```

But trying to do this with the Tables.partitioner, I let it run for ~5 min and it still wasn't done

```
julia> Arrow.write("test2.arrow", Tables.partitioner(profiles) do profile
           CSV.File(profile, datarow=5, header=["clade", "NCBI_taxid", "abundance", "additional_species"])
       end) # hangs...
```

Don't know if it's relevant, but the file after writing from the DataFrame is ~39 Mb, but the partioner one quickly gets to 2.9 Mb and then doesn't get bigger

If I do `Tables.partitioner(first(profiles, 100))`, it completes, making a file that's 2.4 Mb, but doing `first(profiles, 200)`, it does not.

**Comment by quinnj** [comment]
Going to close as I believe this was resolved in the linked PRs

**Comment by kjanosz** [comment]
@quinnj Looks like I have similar issue and it stills occurs with newest Arrow v1.2.4 and Tables v1.4.0 (and CSV v0.8.4 and Julia 1.5.4) versions.
Following snippet simply hangs (CPU usage drops too) after writing around 345MB of data (for three times I checked it, every times it has written the same amount of bytes):
```
r = ZipFile.Reader(file)
data = Tables.partitioner([CSV.File(f, threaded=false) for f in r.files if endswith(f.name, ".txt")]) do f
    f |> DataFrame
end
Arrow.write("data/daily.arrow", data)
close(r)
```


Alternative using @distributed, that loads though entire DataFrame to memory, works (writes around 1.66GB)
```
r = ZipFile.Reader(file)
data = @distributed vcat for f in  [CSV.File(f, threaded=false) for f in r.files if endswith(f.name, ".txt")]
    f |> DataFrame
end
close(r)
Arrow.write("data/daily.arrow", data)
```
Dataset is btw zip taken from [this](https://stooq.com/db/h/) site (daily, us). It's quite a lot of small CSV files, i.e. each stock ticker's historical data is a single file.

**Comment by quinnj** [comment]
Thanks for hte example @kjanosz; I can reproduce and I've been trying to improve the arrow code, but man, ZipFile.Reader is quite a pain to work with. I very quickly run into [this issue](https://github.com/fhs/ZipFile.jl/issues/14) and it also seems like anything multithreaded just really kills reading files from the zip archive. I'll try to dig in further, but if you want a work-around, I'd suggest doing a pre-process loop where you unzip each file and write it out to disk in full, then do the `Tables.partitioner` w/ CSV over the unzipped files. I think that should work currently.

**Comment by quinnj** [comment]
@kjanosz, ok, managed to track down most of the issues on the Arrow.jl side at least. As mentioned in [my PR](https://github.com/JuliaData/Arrow.jl/pull/154), ZipFile seems to not allow non-main-threads to read its files at all, otherwise you get lots of weird reading errors/issues. Anyway, I've added an option to do `ntasks=1` for `Arrow.write` which will avoid any multithreading when writing. I also ran into an issue with PooledArrays getting dict encoded in the arrow data, but over 18K files, we eventually overflow the Int8 encoding type. So with the following code, I'm able to successfully process all the daily ticker data files into a single arrow file:

```julia
using ZipFile, CSV, Tables, Arrow
file = "/Users/jacobquinn/Downloads/d_hk_txt.zip"
r = ZipFile.Reader(file);
data = Tables.partitioner(f->CSV.File(f; pool=false), [f for f in r.files if endswith(f.name, ".txt") && f.uncompressedsize > 0])
Arrow.write("daily.arrow", data; ntasks=1)
```
(we pass `pool=false` to `CSV.File` to tell it to not do any pooling)
